modules/core/src/smithy4s/http/internals/HttpResponseCodeSchemaVisitor.scala

```diff
@@ -89,18 +89,36 @@ class HttpResponseCodeSchemaVisitor()
     fields.flatMap(f => compileField(f)).headOption.getOrElse(NoResponseCode)
   }
 
+  override def union[U](
+      shapeId: ShapeId,
+      hints: Hints,
+      alternatives: Vector[Alt[U, _]],
+      dispatch: Alt.Dispatcher[U]
+  ): ResponseCodeExtractor[U] = {
+    import HttpResponseCodeSchemaVisitor.VariantResponseCodeExtractor
+    val variants = alternatives.map { alt: Alt[U, _] =>
+      VariantResponseCodeExtractor.fromAlt(this, alt)
+    }
+    val allAreNoResponseCode = variants.forall { _.extractor == NoResponseCode }
+    if (allAreNoResponseCode) {
+      NoResponseCode
+    } else {
+      OptionalResponseCode { (u: U) =>
+        variants.flatMap { case variantExtractor =>
+          variantExtractor.project(u).flatMap(variantExtractor.extractor.apply)
+        }.headOption
+      }
```

Comment:
**thought:** I went for `OptionalResponseCode` althought there are other things we could like return `RequiredResponseCode` if _all_ variants are also required. Not sure what would be the difference.  Also could have implemented the `NoResponseCode` by returning an `OptionalResponseCode` that returns `None`